aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,907 @@
+) -> int:
+    """
+    Migrate all the seqr entries from hail batch,
+    and insert them into the aggregate table.
+
+    Break them down by dataset, and then proportion the rest of the costs.
+    """
+
+    # pylint: disable=too-many-locals
+    token = get_hail_token()
+    result = 0
+    lp = f'{log_prefix} ::' if log_prefix else ''
+
+    batches = await get_finished_batches_for_date(
+        start=start, end=end, token=token, billing_project=billing_project
+    )
+    if len(batches) == 0:
+        return 0
+
+    chnk_counter = 0
+    nchnks = math.ceil(len(batches) / entry_chunk_size) * batch_group_chunk_size
+
+    # Process chunks of batches to avoid loading too many entries into memory
+    for btch_grp in chunk(batches, entry_chunk_size):
+        jobs_in_batch = []
+        entries = []
+
+        # Get jobs for a fraction of each chunked batches
+        # to avoid hitting hail batch too much
+        for batch_group_group in chunk(btch_grp, batch_group_chunk_size):
+            chnk_counter += 1
+            times = [b['time_created'] for b in batch_group_group]
+            min_batch = min(times)
+            max_batch = max(times)
+
+            if len(batches) > 100:
+                logger.debug(
+                    f'{lp}Getting jobs for batch chunk {chnk_counter}/{nchnks} '
+                    f'[{min_batch}, {max_batch}]'
+                )
+
+            promises = [get_jobs_for_batch(b['id'], token) for b in batch_group_group]
+            jobs_in_batch.extend(await asyncio.gather(*promises))
+
+        # insert all entries for this batch
+        for batch, jobs in zip(btch_grp, jobs_in_batch):
+            batch['jobs'] = jobs
+            if len(jobs) > 10000 and len(entries) > 1000:
+                logger.info(
+                    f'Expecting large number of jobs ({len(jobs)}) from '
+                    f'batch {batch["id"]}, inserting contents early'
+                )
+                result += insert_new_rows_in_table(
+                    table=GCP_AGGREGATE_DEST_TABLE, objs=entries, dry_run=dry_run
+                )
+                entries = []
+
+            entries_for_batch = func_get_finalised_entries_for_batch(batch)
+            entries.extend(entries_for_batch)
+
+            s = sum(sys.getsizeof(e) for e in entries) / 1024 / 1024
+            if s > 10:
+                logger.info(f'Size of entries: {s} MB, inserting early')
+                result += insert_new_rows_in_table(
+                    table=GCP_AGGREGATE_DEST_TABLE, objs=entries, dry_run=dry_run
+                )
+                entries = []
+
+        result += insert_new_rows_in_table(
+            table=GCP_AGGREGATE_DEST_TABLE, objs=entries, dry_run=dry_run
+        )
+
+    return result
+
+
+RE_matcher = re.compile(r'-\d+$')
+
+
+def billing_row_to_topic(row, dataset_to_gcp_map: dict) -> str | None:
+    """Convert a billing row to a dataset"""
+    try:
+        project_id = row['project']['id']
+    except TypeError:
```

Comment:
Why is this expected behavior?

Reply:
I've changed this code to just type check, I think if row['project'] was None, then this would catch that error, kind of the Pythonic EAFP.